### ISSUE TITLE
[Reproducibility issue] BatchNorm is causing validation accuracy to decrease on loaded model

### ISSUE THREAD

**Issue by AdityaKane2001** [AdityaKane2001/regnets_trainer#8]
@sayakpaul 

### **TLDR**

BatchNorm layers are using the learned mean and var during training, as opposed to mean and var of the batch being processed. This results in the loaded model having lower accuracy than what we observed on training, on WandB dashboard.

### **Long version**

I was trying to reproduce the performance of our stored weights from our best run. However, when I loaded the weights and ran `model.evaluate()`, I got subpar performance (66.5% as opposed to 67%). I checked the input pipeline, it was identical to the one I used while training the same model. 

Then I searched for this on GH and SO. This issue is common when loading weights. This is completely intentional as per [this](https://www.tensorflow.org/api_docs/python/tf/keras/layers/BatchNormalization#note_that).  

As suggested in one of the issues on GH, I ran `model.fit()` on the validation dataset with **`learning_rate=0`**.  This causes the BN layers in the model to "adapt" to the validation dataset ("adapt" because we are using moving mean and var). When I do this, I observe the accuracy back to normal, or at least a lot closer to it. 

See below. The difference between using and not using `fit` is clear.  

```
# all numbers are on validation dataset.

# SESSION 1
# only evaluate() 
50/50 [==============================] - 55s 414ms/step - loss: 1.3860 - accuracy: 0.6637 - top-5-accuracy: 0.8696

# SESSION 2
# fit() with lr=0 
50/50 [==============================] - 67s 357ms/step - loss: 1.3698 - accuracy: 0.6687 - top-5-accuracy: 0.8738
# evaluate() 
50/50 [==============================] - 48s 385ms/step - loss: 1.2010 - accuracy: 0.7060 - top-5-accuracy: 0.8980
```

This occurs because PyTorch simply does not provide the option to freeze the mean and var (this is by implementation in Keras). PyTorch just re-computes the mean and var for all batches separately.

### Conclusion

We are yet again faced with a situation where the difference of implementation in Keras and Pytorch leads to different results by a non-negligible margin. I propose we go with the numbers in WandB, not only because they are better, but because that is closer to the original implementation.
